aws/src/main/java/org/apache/iceberg/aws/ManagedHttpClientRegistry.java

```diff
@@ -0,0 +1,201 @@
+import software.amazon.awssdk.http.SdkHttpClient;
+
+/**
+ * A registry that manages the lifecycle of shared HTTP clients for AWS SDK v2 using reference
+ * counting.
+ */
+public final class ManagedHttpClientRegistry {
+  private static final Logger LOG = LoggerFactory.getLogger(ManagedHttpClientRegistry.class);
+
+  private final ConcurrentMap<String, ManagedHttpClient> clientMap;
+
+  private static volatile ManagedHttpClientRegistry instance;
+
+  public static ManagedHttpClientRegistry getInstance() {
+    if (instance == null) {
+      synchronized (ManagedHttpClientRegistry.class) {
+        if (instance == null) {
+          instance = new ManagedHttpClientRegistry();
+        }
+      }
+    }
+    return instance;
+  }
+
+  private ManagedHttpClientRegistry() {
+    this.clientMap = Maps.newConcurrentMap();
+  }
+
+  /**
+   * Get or create a managed HTTP client for the given configuration. Each call increments the
+   * reference count for the client and returns a ref-counted wrapper.
+   *
+   * @param clientKey unique key identifying the client configuration
+   * @param clientFactory factory to create the HTTP client if not cached
+   * @return a ref-counted HTTP client wrapper
+   */
+  public SdkHttpClient getOrCreateClient(String clientKey, Supplier<SdkHttpClient> clientFactory) {
+    ManagedHttpClient managedClient =
+        clientMap.computeIfAbsent(
+            clientKey,
+            k -> {
+              LOG.debug("Creating new managed HTTP client for key: {}", k);
+              SdkHttpClient httpClient = clientFactory.get();
+              return new ManagedHttpClient(httpClient, k);
+            });
+    // Return the cached ref-counted wrapper
+    return managedClient.acquire();
+  }
+
+  /**
+   * Release a reference to the HTTP client. When the reference count reaches zero, the client is
+   * closed and removed from the registry.
+   *
+   * @param clientKey the key identifying the client to release
+   */
+  public void releaseClient(String clientKey) {
+    ManagedHttpClient managedClient = clientMap.get(clientKey);
+    if (managedClient != null) {
+      if (managedClient.release()) {
+        // Client was closed, remove from map
+        clientMap.remove(clientKey, managedClient);
+      }
+    }
+  }
+
+  @VisibleForTesting
+  ConcurrentMap<String, ManagedHttpClient> clientMap() {
+    return clientMap;
+  }
+
+  @VisibleForTesting
+  void shutdown() {
+    clientMap.values().forEach(ManagedHttpClient::forceClose);
+    clientMap.clear();
+  }
+
+  /**
+   * Managed HTTP client wrapper that provides reference counting for lifecycle management. The HTTP
+   * client is closed when the reference count reaches zero.
+   */
+  static class ManagedHttpClient {
+    private final SdkHttpClient httpClient;
+    private final String clientKey;
+    private final AtomicInteger refCount = new AtomicInteger(0);
+    private volatile boolean closed = false;
+    private volatile WrappedSdkHttpClient wrapper;
```

Comment:
I don't think the wrapper needs to be volatile. Also it can be final

Reply:
Done.